logstash-core/lib/logstash/util/wrapped_synchronous_queue.rb

```diff
@@ -37,5 +37,231 @@ def take
     def poll(millis)
       @queue.poll(millis, TimeUnit::MILLISECONDS)
     end
+
+    def write_client
+      WriteClient.new(self)
+    end
+
+    def read_client
+      ReadClient.new(self)
+    end
+
+    class ReadClient
+      # We generally only want one thread at a time able to access pop/take/poll operations
+      # from this queue. We also depend on this to be able to block consumers while we snapshot
+      # in-flight buffers
+
+      def initialize(queue)
+        @queue = queue
+        @mutex = Mutex.new
+        @inflight_batches = {}
+        @batch_size = 125
+        @wait_for = 5
+      end
+
+      def set_batch_details(batch_size, wait_for)
+        @batch_size = batch_size
+        @wait_for = wait_for
+      end
+
+      def set_events_metric(metric)
+        @event_metric = metric
+      end
+
+      def set_pipeline_metric(metric)
+        @pipeline_metric = metric
+      end
+
+      def inflight_batches
+        @mutex.synchronize do
+          yield(@inflight_batches)
+        end
+      end
+
+      def current_inflight_batch
+        @inflight_batches.fetch(Thread.current, [])
+      end
+
+      def take_batch
+        @mutex.synchronize do
+          batch = ReadBatch.new
+          # guaranteed to be a full batch not a partial batch
+          set_current_thread_inflight_batch(batch)
+          signal = false
+          @batch_size.times do |t|
+            event = (t == 0) ? @queue.take : @queue.poll(@wait_for)
+
+            if event.nil?
+              next
+            elsif event == LogStash::SHUTDOWN || event == LogStash::FLUSH
+              # We MUST break here. If a batch consumes two SHUTDOWN events
+              # then another worker may have its SHUTDOWN 'stolen', thus blocking
+              # the pipeline. We should stop doing work after flush as well.
+              signal = event
+              break
+            else
+              batch.dequeue(event)
+            end
+          end
+          add_dequeued_metrics(batch)
+          [batch, signal]
+        end
+      end
+
+      def set_current_thread_inflight_batch(batch)
+        @inflight_batches[Thread.current] = batch
+      end
+
+      def close_batch(batch)
+        @mutex.synchronize do
+          @inflight_batches.delete(Thread.current)
+        end
+      end
+
+      def add_dequeued_metrics(batch)
```

Comment:
As we discussed in the meeting and recorded here for non-attendees.
Metrics like logging is a cross-cutting concern. Taking metrics measurements can and should occur anywhere.

It could be argued that the pipeline should take the metrics at certain points in the batch lifecycle - but I found that it made the pipeline logic much harder to see.
Also, as the client and batch implementation go together, this metric `@event_metric.increment(:in, batch.dequeued_size)` if done in the pipeline, the pipeline would need to know batch specific internals. The PQ Batch object might not have a dequeued size.